scapy/utils.py

```diff
@@ -1365,33 +1365,78 @@ def __init__(self, filename, fdesc, magic):
             "tsresol": 1000000
         }
         self.blocktypes = {
-            1: self.read_block_idb,
-            2: self.read_block_pkt,
-            3: self.read_block_spb,
-            6: self.read_block_epb,
+            1: self._read_block_idb,
+            2: self._read_block_pkt,
+            3: self._read_block_spb,
+            6: self._read_block_epb,
         }
+        self.endian = "!"  # Will be overwritten by first SHB
+
         if magic != b"\x0a\x0d\x0d\x0a":  # PcapNg:
             raise Scapy_Exception(
                 "Not a pcapng capture file (bad magic: %r)" % magic
             )
-        # see https://github.com/pcapng/pcapng
-        blocklen_, magic = self.f.read(4), self.f.read(4)  # noqa: F841
-        if magic == b"\x1a\x2b\x3c\x4d":
+
+        try:
+            self._read_block_shb()
+        except EOFError:
+            raise Scapy_Exception(
+                "The first SHB of the pcapng file is malformed !"
+            )
+
+    def _read_block(self, size=MTU):
+        # type: (int) -> Optional[Tuple[bytes, RawPcapNgReader.PacketMetadata]]  # noqa: E501
+        try:
+            blocktype = struct.unpack(self.endian + "I", self.f.read(4))[0]
```

Comment:
To my understanding because this is not the first `read` call, this will not raise an error (but will return an empty string if EOF is reached). If `read` returns an empty string, it will trigger a `struct.error` error which is catched.